### PR TITLE
Host implementation in TS

### PR DESCRIPTION
**Pull request by pelikhan**
Allows to mount devices with virtual services. All proto tests passing - pipes. (for later)

### CODE REVIEW

File: src/jdom/bus.ts [hunk 1]
```diff
@@ -731,23 +770,6 @@ export class JDBus extends JDNode {
...
-                        resolve(undefined);
-                    }
-                    else {
-                        // the command timed out
-                        this.emit(TIMEOUT)
-                        this.emit(ERROR, "Timeout (" + timeout + "ms)");
-                        resolve(undefined);
-                    }
-                }
-            }, timeout)
```
Comment (mmoskal) [inline src/jdom/bus.ts:750]: you removed the timeout from withTimeout?

File: src/jdom/bus.ts [hunk 2]
```diff
@@ -731,23 +770,6 @@ export class JDBus extends JDNode {
...
             p.then(v => {
                 if (!done) {
                     done = true
```
Comment (mmoskal) [inline src/jdom/bus.ts:775]: btw, it should probably clearTimeout(tid) here

File: src/jdom/devicehost.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        // compute crc and send
+        pkt.sendCoreAsync(this.bus);
+        // send to current bus
+        this.bus.processPacket(pkt);
+    }
+
+    private handlePacket(pkt: Packet) {
+        const devIdMatch = pkt.deviceIdentifier == this.deviceId;
+        if (pkt.requiresAck && devIdMatch) {
+            pkt.requiresAck = false // make sure we only do it once
```
Comment (mmoskal) [inline src/jdom/devicehost.ts:97]: not sure it makes sense here - in the makecode impl it clears the flag that is on the frame, not packet, but here we're probably not sharing memory

File: src/jdom/devicehost.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            const ack = Packet.onlyHeader(crc)
+            ack.serviceIndex = JD_SERVICE_INDEX_CRC_ACK;
+            this.sendPacketAsync(ack);
+        }
+
+        if (pkt.isMultiCommand) {
+            if (!pkt.isCommand)
+                return; // only commands supported
+            const multiCommandClass = pkt.serviceClass;
+            const h = this._services.find(s => s.serviceClass == multiCommandClass);
```
Comment (mmoskal) [inline src/jdom/devicehost.ts:108]: I wonder if we should handle multiple instances of the same service here... here, in PXT, and in the firmware

File: src/jdom/servicehost.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        return this._registers.find(reg => reg.identifier === BaseReg.StatusCode);
+    }
+
+    protected addRegister(identifier: number, packFormat: string, defaultValue: any[]) {
+        const reg = new JDRegisterHost(this, identifier, packFormat, defaultValue);
+        this._registers.push(reg);
+        return reg;
+    }
+
+    protected addCommand(identifier: number, handler: (pkt) => void) {
```
Comment (mmoskal) [inline src/jdom/servicehost.ts:39]: type missing on `pkt` ? we should have `noImplicitAny` in tsconfig

File: src/jdom/servicehost.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        pkt.serviceIndex = this.serviceIndex;
+        await this.device.sendPacketAsync(pkt);
+    }
+
+    protected sendEvent(event: number, data?: Uint8Array) {
+        const payload = new Uint8Array(4 + (data ? data.length : 0))
+        setNumber(payload, NumberFormat.UInt32LE, 0, event);
+        if (data)
+            memcpy(payload, 4, data);
+        this.sendPacketAsync(Packet.from(SystemCmd.Event, payload))
```
Comment (mmoskal) [inline src/jdom/servicehost.ts:69]: this is not yet doing the triplicate events (which I need to implement in pxt still)

**Review by mmoskal — COMMENTED** [review]
some more remarks

File: src/jdom/bus.ts (hunk 1 above)
Comment (pelikhan) [reply src/jdom/bus.ts:750]: oh TypeScript is too smart and removed it because it things tid is not used.

Comment (mmoskal) [reply src/jdom/bus.ts:750]: I guess we should use it in clearTimeout(), but anyway removing a call with side effects doesn't sound like a good idea

File: src/jdom/devicehost.ts [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    toString() {
+        return `host ${this.shortId}`;
+    }
+
+    async sendPacketAsync(pkt: Packet) {
+        if (!this._bus) return;
+
+        pkt.deviceIdentifier = this.deviceId;
+        // compute crc and send
+        pkt.sendCoreAsync(this.bus);
```
[the remark this answers is not in the document]
Comment (mmoskal) [reply src/jdom/devicehost.ts:89]: I guess you could store the result promise and return it after this.bus.processPacket

File: src/jdom/devicehost.ts (hunk 4 above)
Comment (pelikhan) [reply src/jdom/devicehost.ts:108]: https://github.com/microsoft/jacdac/issues

File: src/jdom/servicehost.ts (hunk 5 above)
Comment (pelikhan) [reply src/jdom/servicehost.ts:39]: need to find this option in gatsby

Comment (mmoskal) [reply src/jdom/servicehost.ts:39]: you can just add it to main tsconfig.json

```
  "noImplicitAny": true,
    "noImplicitReturns": true,
```

There is probably a few places to fix

**Pull request merged**